darkdetect/_linux_detect.py

```diff
@@ -5,41 +5,49 @@
 #-----------------------------------------------------------------------------
 
 import subprocess
+from typing import Callable, Optional
+
+from .base import BaseListener
+
 
 def theme():
     try:
         #Using the freedesktop specifications for checking dark mode
-        out = subprocess.run(
-            ['gsettings', 'get', 'org.gnome.desktop.interface', 'color-scheme'],
-            capture_output=True)
-        stdout = out.stdout.decode()
+        stdout = subprocess.check_output(['gsettings', 'get', 'org.gnome.desktop.interface', 'color-scheme'])
         #If not found then trying older gtk-theme method
         if len(stdout)<1:
-            out = subprocess.run(
-                ['gsettings', 'get', 'org.gnome.desktop.interface', 'gtk-theme'],
-                capture_output=True)
-            stdout = out.stdout.decode()
-    except Exception:
-        return 'Light'
+            stdout = subprocess.check_output(['gsettings', 'get', 'org.gnome.desktop.interface', 'gtk-theme'])
+    except subprocess.SubprocessError:
+        return "Light"
     # we have a string, now remove start and end quote
-    theme = stdout.lower().strip()[1:-1]
-    if '-dark' in theme.lower():
-        return 'Dark'
-    else:
-        return 'Light'
-
-def isDark():
-    return theme() == 'Dark'
-
-def isLight():
-    return theme() == 'Light'
-
-# def listener(callback: typing.Callable[[str], None]) -> None:
-def listener(callback):
-    with subprocess.Popen(
-        ('gsettings', 'monitor', 'org.gnome.desktop.interface', 'gtk-theme'),
-        stdout=subprocess.PIPE,
-        universal_newlines=True,
-    ) as p:
-        for line in p.stdout:
-            callback('Dark' if '-dark' in line.strip().removeprefix("gtk-theme: '").removesuffix("'").lower() else 'Light')
+    theme_: bytes = stdout.lower().strip()[1:-1]
+    return "Dark" if b"-dark" in theme_.lower() else "Light"
+
+
+class GnomeListener(BaseListener):
+    """
+    A listener for Gnome on Linux
+    """
+
+    def __init__(self, callback: Callable[[str], None]):
+        self._proc: Optional[subprocess.Popen] = None
```

Comment:
I don't think you have to initialize this, in this way you can avoid the `Optional`. Suggested change:
```python
self._proc: subprocess.Popen
```

This applies also to the macOS listener.

Reply:
I have no issues with either; I did this because `pylint` was complaining about defining a variable outside of `__init__`. Which would you prefer?

Reply:
In this case, I would definitely prefer to have the `pylint` warning.

Reply:
Actually, I realized my initial code was incomplete. `Optional` should be used and the other processes should check if it exists to allow for a user calling `.stop()` or `.wait()` before `.listen()`; which I will have raise `RuntimeError`s.

Reply:
I am confused, I thought `self._state` was meant for this? I am not sure I would use `self._proc` for this checking, as it would force you to re-implement this "safety" logic in all the OS modules. Conversely, this check should be in the base class (where it is now).

Reply:
Hmm, I was being extra cautions; perhaps unnecessarily. `self._state` should indeed handle this. I was thinking about if someone used internal methods; I should not. I will do what you said initially instead of overcomplicating things. Sorry about the confusion!